rice-datanode/source/data_transfer_server.cc

```diff
@@ -318,7 +324,121 @@ void TransferServer::synchronize(std::function<void(TransferServer&, tcp::socket
 	cv.notify_one();
 }
 
+bool TransferServer::replicate(uint64_t len, std::string ip, std::string xferport, ExtendedBlockProto blockToTarget) {
+
+	LOG(INFO) << " replicating length " << len << " with ip " << ip << " and port " << xferport;
+
+	std::string blk;
+	LOG(INFO) << "blockToTarget is " << blockToTarget.blockid();
+
+    if (fs->getBlock(blockToTarget.blockid(), blk)) {
```

Comment:
I think we should add a hasBlock to native fs so we don't read in the whole block here just to check that it exists.